File: guests/go/milestone2/milestone2.go
// The milestone-2 scene from Go, on the construction sugar.
//
// WHAT THIS SCENE DOCUMENTS IS HOW OCCURRENCES REACH AN APP, and only
// that. The remove button is a STAMPED copy, so its handler is
// registered CENTRALLY against the template node and receives that
// copy's key path; the live step button rides its constructor.
//
// AND THE APP NAMES EVERY KEY HERE, on purpose: "g1" and "a" are this
// app's own identity for a group and an item, re-addressed later by
// name. That is exactly what InsertFresh is NOT for
// (docs/fresh-key-plan.md).
//
//	KAYA_SELFTEST=1 go run dev.kaya/guests/go/cmd
package milestone2

import (
	"fmt"

	kaya "dev.kaya/bindings/go"
)

func App() *kaya.App {
	app := kaya.NewApp()

	// The handles the central registration below needs: a node parents
	// into its container AT CREATION, so both are built inside the
	// template bodies that hold them and escape through these.
	var (
		status       kaya.Signal[string]
		items        kaya.Collection
		removeButton kaya.Node
	)

	steps := 0
	app.Build(func(tx *kaya.Tx) {
		status = tx.Signal("step 0")
		extras := tx.Signal(false)

		groups := tx.Collection()

		tx.Mount(tx.Column(func() {
			tx.Button("step", func(tx *kaya.Tx) {
				steps++
				switch steps {
				case 1:
					tx.Insert(groups, "g1", "Work")
					todos := items.At("g1")
					tx.Insert(todos, "a", "send report")
					tx.Insert(todos, "b", "buy milk")
				case 2:
					tx.Insert(groups, "g2", "Home")
					tx.Insert(items.At("g2"), "a", "water plants")
					tx.Update(groups, "g1", "Office")
				}
				tx.Write(extras, steps == 1)
				tx.Write(status, fmt.Sprintf("step %d", steps))
			})
			tx.Label(status)
			tx.When(extras, func(t *kaya.Tpl) {
				t.LabelText("extras on")
			})
			for group := range tx.Rows(groups).All() {
				group.Column(func() {
					group.Label(group.Value())

					items = group.Collection()
					for item := range group.Rows(items).All() {
						item.Column(func() {
							item.Label(item.Value())
							removeButton = item.Button("remove")
						})
					}
				})
			}
		}))
	})

	app.OnClickNode(removeButton, func(tx *kaya.Tx, keys []any) {
		group, item := keys[0].(string), keys[1].(string)
		todos := items.At(group)
		tx.Remove(todos, item)
		left := tx.Len(todos)
		tx.Write(status, fmt.Sprintf("removed %s/%s, %d left", group, item, left))
	})

	return app
}

File: guests/java/dev/kaya/guests/Milestone2.java
package dev.kaya.guests;

import dev.kaya.KayaApp;

/**
 * The milestone2 scene from the JVM — guests/rust/milestone2.rs,
 * tools/scenes/milestone2.steps.
 */
public final class Milestone2 {
    private static final class Scene {
        final KayaApp.Signal<String> status;
        final KayaApp.Collection items;
        final KayaApp.Node removeButton;

        Scene(KayaApp.Signal<String> status, KayaApp.Collection items,
                KayaApp.Node removeButton) {
            this.status = status;
            this.items = items;
            this.removeButton = removeButton;
        }
    }

    /** Java lambdas cannot assign captured locals. */
    private static final class Refs {
        KayaApp.Collection items;
        KayaApp.Node remove;
    }

    private static int steps;

    public static void app() {
        KayaApp app = new KayaApp();

        Scene scene = app.build(tx -> {
            KayaApp.Signal<String> status = tx.signal("step 0");
            KayaApp.Signal<Boolean> extras = tx.signal(false);

            KayaApp.Collection groups = tx.collection();

            Refs refs = new Refs();

            tx.mount(tx.column(() -> {
                tx.button("step", t -> { // button#0
                    steps++;
                    if (steps == 1) {
                        t.insert(groups, "g1", "Work");
                        KayaApp.Collection todos = refs.items.at("g1");
                        t.insert(todos, "a", "send report");
                        t.insert(todos, "b", "buy milk");
                    } else if (steps == 2) {
                        t.insert(groups, "g2", "Home");
                        t.insert(refs.items.at("g2"), "a", "water plants");
                        t.update(groups, "g1", "Office");
                    }
                    t.write(extras, steps == 1);
                    t.write(status, "step " + steps);
                });
                tx.label(status); // label#0
                // A BLOCK body: an expression lambda is ambiguous between the
                // Consumer and Function `when` overloads (docs/traps.md).
                tx.when(extras, t -> {
                    t.label("extras on");
                });
                for (var group : tx.rows(groups)) {
                    group.column(() -> {
                        group.label(group.value());

                        refs.items = group.collection();
                        for (var item : group.rows(refs.items)) {
                            item.column(() -> {
                                item.label(item.value());
                                refs.remove = item.button("remove");
                            });
                        }
                    });
                }
            }));
            return new Scene(status, refs.items, refs.remove);
        });

        app.onClick(scene.removeButton, (tx, keys) -> {
            String group = (String) keys.get(0);
            String item = (String) keys.get(1);
            KayaApp.Collection todos = scene.items.at(group);
            tx.remove(todos, item);
            int left = tx.count(todos);
            tx.write(scene.status, "removed " + group + "/" + item + ", " + left + " left");
        });

        app.dispatchLoop();
    }

    private Milestone2() {}
}
